lib/tasks_private/lib/aws_instance_data_parser.rb

```diff
@@ -0,0 +1,316 @@
+    end
+  end.freeze
+
+  # for :description
+  POPULAR_TYPES = %w(t1 t2).freeze
+
+  # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/
+  #   using-vpc.html#vpc-only-instance-types
+  VPC_ONLY_TYPES = %w(m4 m5 t2 c4 c5 r4 x1 h1 i3 f1 g3 p2 p3).freeze
+
+  # https://github.com/ManageIQ/manageiq-providers-amazon/
+  #   blob/933a3d08e0adb012c7cbefbaeaa262a81c855fe1/
+  #   lib/tasks_private/instance_types.rake#L49
+  CLUSTERABLE_TYPES = %w(m4 c3 c4 cr1 r4 x1 hs1 i2 g2 p2 d2).freeze
+
+  # some types data missing this
+  EBS_OPTIMIZED_TYPES = %w(x1).freeze
+
+  # some types data missing 'intel' features info
+  CPU_FEATURES = {
+    'Intel Xeon E5-2670'                              => %i(avx turbo).freeze,
+    'Intel Xeon E5-2670 v2 (Ivy Bridge/Sandy Bridge)' => %i(avx turbo).freeze,
+    'Intel Xeon E5-2666 v3 (Haswell)'                 => %i(avx avx2 turbo).freeze,
+    'Intel Xeon E5-2686 v4 (Broadwell)'               => %i(avx avx2 turbo).freeze,
+  }.freeze
+
+  # considering all types have it except of legacy ones
+  NON_AES_NI_TYPES = %w(t1 m1).freeze
+
+  # some types data missing this (here is only those who missing)
+  # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/
+  #   enhanced-networking.html#supported_instances
+  ENHANCED_NETWORKING_TYPES = %w(f1).freeze
+
+  ParsedName = Struct.new(:base_type, :size_factor, :size_name)
+  ParsedStorage = Struct.new(:volumes, :size, :type)
+
+  private_constant(*constants(false).without(:REQUIRED_ATTRIBUTES))
+
+  attr_reader :product_data
+
+  def initialize(product_data)
+    @parsed = false
+    @product_data = product_data
+    @unknown_values = {}
+  end
+
+  def result
+    [instance_data, unknown_values]
+  end
+
+  def instance_data
+    parse! unless @parsed
+    @instance_data
+  end
+
+  def unknown_values
+    parse! unless @parsed
+    @unknown_values
+  end
+
+  ### individual attributes
+
+  ## general
+
+  def current_generation?
+    product_data['currentGeneration'] == 'Yes'
+  end
+
+  def current_version?
+    !!product_data['currentVersion']
+  end
+
+  def instance_type
+    product_data['instanceType']
+  end
+
+  def instance_family
+    product_data['instanceFamily']
+  end
+
+  delegate :base_type, :to => :parsed_name
+
+  def description
+    description = [base_type.upcase]
+    description << instance_family unless POPULAR_TYPES.include?(base_type)
+    description << (size_name == 'xlarge' ? "#{size_factor}XL" : size_name.capitalize)
+    description.join(' ')
+  end
+
+  ## virtualization
+
+  def virtualization_type
+    VIRT_TYPES[base_type]
+  end
+
+  delegate :size_factor, :size_name, :to => :parsed_name
+
+  def vcpus
+    product_data['vcpu'].to_i
+  end
+
+  ## cpu
+
+  def physical_processor
+    product_data['physicalProcessor']
+  end
+
+  def cpu_clock_speed
+    product_data['clockSpeed']
+  end
+
+  def cpu_arches
+    CPU_ARCHES.fetch(product_data['processorArchitecture']) do |cpu_arch|
```

Comment:
I found a simpler way to do this....

```ruby
CPU_ARCHES => {32 => "i386", 64 => "x86_64"}.freeze

product_data['processorArchitecture'].scan(/\d+/).map { |bits| CPU_ARCHES[bits] }
```

Reply:
This way we will not be able to control if that field contains something unexpected.

Reply:
So you prefer the hardcoded-ness of it?

Reply:
Yes, this way we will not be blind when Amazon'll introduce something new here.